components/brave_wallet/browser/zcash/zcash_orchard_storage.cc

```diff
@@ -418,4 +586,864 @@ std::optional<ZCashOrchardStorage::Error> ZCashOrchardStorage::UpdateNotes(
+        Error{ErrorCode::kDbInitError, database_.GetErrorMessage()});
+  }
+
+  sql::Statement resolve_shard_statement(database_.GetCachedStatement(
+      SQL_FROM_HERE, "SELECT root_hash, shard_data FROM " kShardTree " "
+                     "WHERE account_id = ? AND shard_index = ?;"));
+
+  resolve_shard_statement.BindString(0, account_id->unique_key);
+  resolve_shard_statement.BindInt64(1, address.index);
+
+  if (!resolve_shard_statement.Step()) {
+    if (!resolve_shard_statement.Succeeded()) {
+      return base::unexpected(Error{ErrorCode::kFailedToExecuteStatement,
+                                    database_.GetErrorMessage()});
+    }
+    return base::ok(std::nullopt);
+  }
+
+  auto hash = ReadRootHash(resolve_shard_statement, 0);
+  if (!hash.has_value()) {
+    return base::unexpected(Error{ErrorCode::kDbInitError, hash.error()});
+  }
+
+  auto shard = OrchardShard(address, hash.value(), std::vector<uint8_t>());
+
+  shard.shard_data = base::ToVector(resolve_shard_statement.ColumnBlob(1));
+
+  return base::ok(std::move(shard));
+}
+
+base::expected<std::optional<OrchardShard>, ZCashOrchardStorage::Error>
+ZCashOrchardStorage::LastShard(const mojom::AccountIdPtr& account_id,
+                               uint8_t shard_height) {
+  auto shard_index = GetLatestShardIndex(account_id);
+  if (!shard_index.has_value()) {
+    return base::unexpected(shard_index.error());
+  }
+
+  if (!shard_index.value()) {
+    return base::ok(std::nullopt);
+  }
+
+  return GetShard(account_id, OrchardShardAddress{shard_height,
+                                                  shard_index.value().value()});
+}
+
+base::expected<std::vector<OrchardShardAddress>, ZCashOrchardStorage::Error>
+ZCashOrchardStorage::GetShardRoots(const mojom::AccountIdPtr& account_id,
+                                   uint8_t shard_level) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  if (!EnsureDbInit()) {
+    return base::unexpected(
+        Error{ErrorCode::kDbInitError, database_.GetErrorMessage()});
+  }
+
+  std::vector<OrchardShardAddress> result;
+
+  sql::Statement resolve_shards_statement(database_.GetCachedStatement(
+      SQL_FROM_HERE, "SELECT shard_index FROM " kShardTree
+                     " WHERE account_id = ? ORDER BY shard_index;"));
+
+  resolve_shards_statement.BindString(0, account_id->unique_key);
+
+  while (resolve_shards_statement.Step()) {
+    auto shard_index = ReadUint32(resolve_shards_statement, 0);
+    if (!shard_index) {
+      return base::unexpected(Error{ErrorCode::kFailedToExecuteStatement, ""});
+    }
+    result.push_back(OrchardShardAddress{shard_level, shard_index.value()});
+  }
+
+  if (!resolve_shards_statement.is_valid()) {
+    return base::unexpected(Error{ErrorCode::kFailedToExecuteStatement, ""});
+  }
+
+  return base::ok(std::move(result));
+}
+
+base::expected<bool, ZCashOrchardStorage::Error>
+ZCashOrchardStorage::AddCheckpoint(const mojom::AccountIdPtr& account_id,
```

Comment:
nit: the complexity of this function is pretty large. It might make sense to split this up a bit further in a refactor or other PR where you're touching this code later.